domFor: always get generation from delayedRemoval instead of parameter

kfule opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
The `generation` of domFor is no longer passed as a parameter.
This allows domFor to work well in onbeforeremove and onremove and reduces the amount of code.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
#3005 did not properly take into account that domFor is public API.
This PR fixes these problems (https://github.com/MithrilJS/mithril.js/pull/3005#issuecomment-2628779554):
- The domFor in onremove() does not work well during delayed removal. ([flems](https://flems.io/#0=N4IgtglgJlA2CmIBcBWADAOjQDgDQgDMIEBnZAbVADsBDMRJEDAKzPwGMB7KgF3l+SEArlXY8I3AARR4sGgE8AFGBIBKScAA6VTTwBO8HkL1VJVeAHdJABT2dIJeIoMlOsAG7xJAXgB8GyUceABUIek4hHmd4Vw94XEkVdQBfVW1k7W0EHkl3CBIIACMEH0l9IXhtMAwwCN5FKE52IXpeDELOKHkE4DzLJElFdT9Jcm1JXPyikoAyGcTFTRAoCHclnvGJyW5C+AJOAwNazwHFdypO+GH-LVMtia4qWPgMWE4Ac0WQHb2D+CPOCclmk7vd9npBu4aBDGmBtgREhhYQAxA5nC4yVTqW73e6PZ6vD4NezqAD0pMkAAc9BBePAoJt7hlQVsDEYTNJZApFABGND8kFM3CM7ZUAEnSEYq4+G4ih7cAlvT5Lbji+BIYFyyTgyHQ6T2eGIlFo86XLEaLXyp5uF5K4lgMkUi45am0vgMlkTZlMzbJBJLfacYHCu5gL6FSI8bjrDTcdiwCDsADWp2uGjyBWKXm8kgAhBnpvBkn7JEtKUISAALTVUAC6qW0IHwjgQYgkT0EACYAMxINAgZK1-AJqhJshISggWj0QSQHiVmmwJsgYxLxiVng8SkkJDkkSUpPvDBcMCkucL4gAAU7GBvPO7Z4g88XNVpLDYIB48kpDBAJHYNKUjwA71kAA))
- When there is onbeforeremove() in both vnode.state and vnode.attrs, domFor in onbeforeremove() in the latter vnode.attrs does not work well (since 3005.) ([flems](https://flems.io/#0=N4IgtglgJlA2CmIBcBWADAOjQDgDQgDMIEBnZAbVADsBDMRJEDAKzPwGMB7KgF3l+SEArlXY8I3AARR4sGgE8AFGBIBKScAA6VTTwBO8HkL1VJVeAHdJABT2dIJeIoMlOsAG7xJAXgB8GyUceABUIek4hHmd4Vw94XEkVdQBfVW1k7W0uKhIeSQBhewAHHw1MnncISyRJRXU-RMVNECgId2aE5oJOTmbVXHLuACN4boMDME5PGsV3Kk4Zev8tU0k1gHoAKm01tbHa9xo9aXtJTgJEjCh7ADFOPVn5xfUV3d3s2PgMWE4Ac0VrmB1Ot1pIinoILx4FAdrsMqs1pt1rC1gYjCZpLIFIoAIxofFpVbw+HaBB5SokCBDBClfRCeDaMAYSYiKLXdhCei8DBDBbyBLASrVWpLSTkWEUqk0gBk0sahTARQFKLOVBGY3gEym8BmcwW8FFrzekn2syOJzAZwuTMBdweeueZQRxo+bi+P3+gOBoPmeXBkL4MOda3hxskaOMphkciUeIJKtDklSA1WYCaICGkR43A6Gm47FgEHYAGsZobJdSvN5JABCCsIZLJTogIpCEgACz62gAuqltCB8I4EGIJDlBAAmbBINAgZLd-CFqjFshISggWj0QSQHjtiGwAACAGYsFgALTmAAePAwADYAJwDkDGWCCds8HhFEhIEEiIrF34YFwYDrNuu7EEeJ5oOe8BXred4gRAO57sykIsGwIA8PIRQMCAJDsBCRQ8LOvZAA))

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
`npm run test` including new tests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] My change requires a documentation update, and I've opened a pull request to update it already:
- [x] I have read https://mithril.js.org/contributing.html.

dead-claudia reviewed (changes requested):
I'd like to see tests confirming multiple generations are tracked correctly, for completeness.

Given you do this:

- Render vnode A with `onbeforeremove` and DOM nodes A1 and A2
- Remove vnode A, do not resolve `onbeforeremove`
- Render vnode B with `onbeforeremove` and DOM nodes B1 and B2
- Remove vnode B, do not resolve `onbeforeremove`
- Render vnode C with `onbeforeremove` and DOM nodes C1 and C2
- Remove vnode C, do not resolve `onbeforeremove`

I'd like to see confirmation of this (6 possible removal sequences in total):

- `m.domFor(A)` returns `[A1, A2]` consistently before A's `onbeforeremove` resolves, even if B's and/or C's resolves first
- `m.domFor(B)` returns `[B1, B2]` consistently before B's `onbeforeremove` resolves, even if A's and/or C's resolves first
- `m.domFor(C)` returns `[C1, C2]` consistently before C's `onbeforeremove` resolves, even if B's and/or C's resolves first

kfule commented:
@dead-claudia Thanks for your review comments.
I have added the following tests as you commented.
- `m.domFor(A)` returns `[A1, A2]` consistently before A's `onbeforeremove` resolves, even if B's resolves first
- `m.domFor(A)` returns `[A1, A2]` consistently before A's `onbeforeremove` resolves, even if B's and C's resolves first
- `m.domFor(B)` returns `[B1, B2]` consistently before B's `onbeforeremove` resolves, even if C's resolves first
- `m.domFor(B)` returns `[B1, B2]` consistently before B's `onbeforeremove` resolves, even if C's and A's resolves first
- `m.domFor(C)` returns `[C1, C2]` consistently before C's `onbeforeremove` resolves, even if A's resolves first
- `m.domFor(C)` returns `[C1, C2]` consistently before C's `onbeforeremove` resolves, even if A's and B's resolves first

I am a little sorry that the test codes are somewhat verbose.
It seems that the test codes could be more aggregated, but I did not want to make it too complicated.

dead-claudia commented:
> I am a little sorry that the test codes are somewhat verbose.
> It seems that the test codes could be more aggregated, but I did not want to make it too complicated.

@kfule Oh, I expected it to be messy. Don't worry.

dead-claudia reviewed (changes requested):
You got these orders:

- BCA
- CAB
- ABC

I also wanted to see the other three:

- ACB
- BAC
- CBA

kfule commented:
@dead-claudia 
I have added the test as you commented (ACB, BAC and CBA.)
I also removed the `generation` parameter of `removeDOM()` that I forgot to remove.
